students/ZidanLuo/session09/donor_models.py

```diff
@@ -0,0 +1,193 @@
+
+    
+    def add_donor(self, donor_name, donation_amount):
+        """
+        Add a donor to the donor collection, and compare the length of the donor's name and donation 
+        amount to decide the widths we are going to use in generating reports
+
+        :param: name of the donor and donation amount
+        :return: the new donor database
+        """
+        if donor_name in self.donor_dict:
+            self.donor_dict[donor_name].add_donation(donation_amount)
+        else:
+            self.donor_dict[donor_name] = Donor(donor_name, donation_amount)
+
+        if len(donor_name) > len(self.longest_name):
+            self.longest_name = donor_name
+
+        if self.donor_dict[donor_name].sum_donations > self.largest_amount:
+            self.largest_amount = self.donor_dict[donor_name].sum_donations
+
+    
+    def sort_report(self):
+        """
+        sort dictionary by the total donation amount donated by each donor
+
+        :param: 
+        :return: a list of tuple of sorted donor
+        """
+        self.sorted_dict = sorted(
+            self.donor_dict.items(),
+            key=lambda x: x[1].sum_donations,
+            reverse=True)
+
+    
+    def send_single(self, donor, last_donation_amount):
+        """
+        find the donor in the dictionary and get the formatted thank you letter
+
+        :param: name of the donor and donation amount
+        :return: a string type formatted thank you letter
+        """
+        return self.donor_dict[donor].letter_generator(last_donation_amount)
+
+    
+    def report_generator(self):
+        """
+        Generate a report of the donors donation history
+
+        :return: a string type report
+        """
+        report = []
+        width = len(self.longest_name) + 8
+        width2 = len(str(self.largest_amount)) + 8
+
+        self.sort_report()
+
+        report.append("{:{width}}{}{:^{width2}}{}{:^10}{}{:^{width2}}{}\n".format(
+            "      Donor Name",
+            "|",
+            " Total Given",
+            "|",
+            " Num Gifts",
+            "|",
+            "Average Gift",
+            "|",
+            width=width,
+            width2=width2))
+        report.append("-" * (width + width2 * 2 + 16) + "\n")
+        
+        [report.append("{:{width}}{:1}{:{width2}.2f}{:>11}{:>2}{:>{width2}.2f}\n".format(
+            x[0],
+            "$",
+            x[1].sum_donations,
+            x[1].num_donations,
+            " $",
+            round(x[1].average_donation, 2),
+            width=width,
+            width2=width2)) for x in self.sorted_dict]
+
+        return "".join(report)
+
+    
+    def save_all(self, out_path):
+        """
+        evoke save file function in donor class for each donor
+
+        :param: string of output path
+        :return: 
+        """
+        [donor.save_to_disk(out_path) for donor in self.donor_dict.values()]
+
+    
+    def get_longest_name(self):
```

Comment:
NEVER write getters like this in Python! these are all attributes, you can simply access them:

`the_db.longest_name`

Then, if you ever do need to add some logic, you can make them properties.